frame/staking/src/lib.rs

```diff
@@ -1438,18 +1679,176 @@ decl_module! {
+		/// m`.
+		///
+		/// major steps (all done in `check_and_replace_solution`):
+		///
+		/// - Storage: O(1) read `ElectionStatus`.
+		/// - Storage: O(1) read `PhragmenScore`.
+		/// - Storage: O(1) read `ValidatorCount`.
+		/// - Storage: O(1) length read from `SnapshotValidators`.
+		/// - Storage: O(v) reads of `AccountId`.
+		/// - Memory: O(m) iterations to map winner index to validator id.
+		/// - Storage: O(n) reads `AccountId`.
+		/// - Memory: O(n + m) reads to map index to `AccountId` for un-compact.
+		/// - Storage: O(e) accountid reads from `Nomination` to read correct nominations.
+		/// - Storage: O(e) calls into `slashable_balance_of_extended` to convert ratio to staked.
+		/// - Memory: build_support_map. O(e).
+		/// - Memory: evaluate_support: O(E).
+		/// - Storage: O(e) writes to `QueuedElected`.
+		/// - Storage: O(1) write to `QueuedScore`
+		///
+		/// The weight of this call is 1/10th of the blocks total weight.
+		/// # </weight>
+		#[weight = SimpleDispatchInfo::FixedNormal(100_000_000)]
+		pub fn submit_election_solution(
+			origin,
+			winners: Vec<ValidatorIndex>,
+			compact_assignments: CompactAssignments,
+			score: PhragmenScore,
+		) {
+			let _who = ensure_signed(origin)?;
+			Self::check_and_replace_solution(
+				winners,
+				compact_assignments,
+				ElectionCompute::Signed,
+				score,
+			)?
+		}
+
+		/// Unsigned version of `submit_election_solution`. Will only be accepted from those who are
+		/// in the current validator set.
+		#[weight = SimpleDispatchInfo::FixedNormal(100_000_000)]
+		pub fn submit_election_solution_unsigned(
+			origin,
+			winners: Vec<ValidatorIndex>,
+			compact_assignments: CompactAssignments,
+			score: PhragmenScore,
+			// already used and checked in ValidateUnsigned.
+			_validator_index: u32,
+			_signature: <T::KeyType as RuntimeAppPublic>::Signature,
+		) {
+			ensure_none(origin)?;
+			Self::check_and_replace_solution(
+				winners,
+				compact_assignments,
+				ElectionCompute::Authority,
+				score,
+			)?
+		}
 	}
 }
 
 impl<T: Trait> Module<T> {
-	// PUBLIC IMMUTABLES
-
 	/// The total balance that can be slashed from a stash account as of right now.
 	pub fn slashable_balance_of(stash: &T::AccountId) -> BalanceOf<T> {
 		Self::bonded(stash).and_then(Self::ledger).map(|l| l.active).unwrap_or_default()
 	}
 
-	// MUTABLES (DANGEROUS)
+	/// internal impl of [`slashable_balance_of`] that returns [`ExtendedBalance`].
+	fn slashable_balance_of_extended(stash: &T::AccountId) -> ExtendedBalance {
+		<T::CurrencyToVote as Convert<BalanceOf<T>, u64>>::convert(
+			Self::slashable_balance_of(stash)
+		) as ExtendedBalance
+	}
+
+	/// The session index of the given era.
+	fn start_session_index_of(era: EraIndex) -> SessionIndex {
+		Self::eras_start_session_index(era)
+			.unwrap_or_else(|| {
+				frame_support::print("Error: start_session_index must be set for current_era");
+				0
+			})
+	}
```

Comment:
either you rename this function `fn current_era_start_session_index()` or you document that era must exist or 0 is returned with a warning and you change the error message.

Reply:
I actually will remove this back to not being a function. At first I used it at multiple places but now it is used onyl once.